File: problem3/c#/Problem3.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
/*
# Problem 3

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600,851,475,143 ?
 */

namespace Problem3
{
    public static class PrimeFactor
    {
        public static long LargestPrimeFactor(long num)
        {

            long largest_factor = 1;

            // remove any factors of 2 first
            while (num % 2 == 0)
            {
                largest_factor = 2;
                num = num / 2;
            }

            // now look at odd factors
            long divisor = 3L;
            while (num != 1)
            {
                while (num % divisor == 0)
                {
                    largest_factor = divisor;
                    num = num / divisor;
                }
                divisor += 2;
            }

            return largest_factor;
        }
    }

    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void FindLargestPrimeFactor_Test()
        {
            Assert.AreEqual(2, PrimeFactor.LargestPrimeFactor(2));
            Assert.AreEqual(5, PrimeFactor.LargestPrimeFactor(5));
            Assert.AreEqual(5, PrimeFactor.LargestPrimeFactor(10));
            Assert.AreEqual(29, PrimeFactor.LargestPrimeFactor(13195));
            Assert.AreEqual(6857, PrimeFactor.LargestPrimeFactor(600851475143));
        }
    }
}

File: problem3/python/problem3.py
#!/usr/bin/env python3
import unittest

class PrimeFactorTest(unittest.TestCase):
    
    def setUp(self):
        self.prime_factor = PrimeFactor()
    
    def test_sum_amount_1000(self):
        self.assertEqual(2, self.prime_factor.largest_prime_factor(2))
        self.assertEqual(5, self.prime_factor.largest_prime_factor(5))
        self.assertEqual(5, self.prime_factor.largest_prime_factor(10))
        self.assertEqual(29, self.prime_factor.largest_prime_factor(13195))
        self.assertEqual(6857, self.prime_factor.largest_prime_factor(600851475143))
        
        # The following takes too long as we cannot reduce the original number
        # self.assertEqual(600851475149, self.prime_factor.largest_prime_factor(600851475149))

class PrimeFactor:
    
    def largest_prime_factor(self, num):
        largest_factor = 1
 
        # remove any factors of 2 first
        while num % 2 == 0:
            largest_factor = 2
            num = num // 2
    
        # now look at odd factors
        divisor = 3
        while num != 1:
            while num % divisor == 0:
                largest_factor = divisor
                num = num // divisor
            divisor += 2
    
        return largest_factor

if __name__ == '__main__':
    unittest.main()
